synapse/logging/context.py

```diff
@@ -751,6 +787,9 @@ def run_in_background(f, *args, **kwargs) -> defer.Deferred:
     # At this point we should have a Deferred, if not then f was a synchronous
     # function, wrap it in a Deferred for consistency.
     if not isinstance(res, defer.Deferred):
+        # All `Awaitable`s in Synapse are either coroutines or `Deferred`s
+        assert not isinstance(res, Awaitable)
+
```

Comment:
Not sure about this, but also not sure if I can put my finger on why this gives me a bad feeling.

Reply:
I considered accepting a union, but decided it wasn't worth the hassle of replacing `Awaitable`s throughout synapse.

I'm not sure how we'd handle other awaitables here. Create a dummy async function which `await`s them?

Reply:
Ahh. I don't think I'd fully appreciated that `Awaitable` and `Coroutine` were different concepts.

Would it help to have a Synapse-only type alias for `Union[Coroutine-that-returns-T, Deferred[T]]`?

Reply:
After some discussion in #synapse-dev, the best thing to do seems to be to leave the assert as is.